src/julia_gc.c

```diff
@@ -0,0 +1,641 @@
+{
+  // HOOK: return `size` bytes memory of TNUM `type`.
+  void *result = (void *) jl_gc_alloc(JuliaTLS, size, datatype_bag);
+  return result;
+}
+
+void InitMarkFuncBags (
+    UInt                type,
+    TNumMarkFuncBags    mark_func )
+{
+  // HOOK: set mark function for type `type`.
+}
+
+void InitSweepFuncBags (
+    UInt                type,
+    TNumSweepFuncBags    mark_func )
+{
+  // HOOK: set sweep function for type `type`.
+  // This is intended for weak pointer objects.
+}
+
+static void TryMark(void *p) 
+{
+  jl_value_t *p2 = jl_pool_base_ptr(p);
+  if (!p2)
+    p2 = treap_find(bigvals, p);
+  if (p2) {
+    jl_gc_queue_root(p2);
+  }
+}
+
+void CHANGED_BAG(Bag bag) {
+  // This is the simple Julia write barrier; jl_gc_wb() is more accurate
+  // but requires a pointer to the originating back, too, so we go with
+  // this one for now.
+  jl_gc_wb_back((void *)bag);
+}
+
+void GapRootScanner(int global) {
+  syJmp_buf registers;
+  sySetjmp(registers);
+  for (char *p = (char *)registers; p < (char *) GapStackBottom; p += sizeof(Int)) {
+    TryMark(*(void **)p);
+  }
+  for (Int i = 0; i < GlobalCount; i++) {
+    Bag p = *GlobalAddr[i];
+    if (IS_BAG_REF(p))
+      jl_gc_queue_root((jl_value_t *) p);
+  }
+}
+
+void            InitBags (
+    UInt                initial_size,
+    TNumStackFuncBags   stack_func,
+    Bag *               stack_bottom,
+    UInt                stack_align)
+{
+    // HOOK: initialization happens here.
+    jl_extend_init();
+    jl_gc_enable(0); /// DEBUGGING
+    GapStackBottom = stack_bottom;
+    GapStackAlign = stack_align;
+    JuliaTLS = jl_extend_get_ptls_states();
+    jl_root_scanner_hook = GapRootScanner;
+    jl_nonpool_alloc_hook = alloc_bigval;
+    jl_nonpool_free_hook = free_bigval;
+}
+
+UInt CollectBags (
+    UInt                size,
+    UInt                full )
+{
+    // HOOK: perform a full collection
+    return 1;
+}
+
+void            RetypeBag (
+    Bag                 bag,
+    UInt                new_type )
+{
+    BagHeader * header = BAG_HEADER(bag);
+
+#ifdef COUNT_BAGS
+    /* update the statistics      */
+    {
+          UInt                old_type;       /* old type of the bag */
+          UInt                size;
+
+          old_type = header->type;
+          size = header->size;
+          InfoBags[old_type].nrLive   -= 1;
+          InfoBags[new_type].nrLive   += 1;
+          InfoBags[old_type].nrAll    -= 1;
+          InfoBags[new_type].nrAll    += 1;
+          InfoBags[old_type].sizeLive -= size;
+          InfoBags[new_type].sizeLive += size;
+          InfoBags[old_type].sizeAll  -= size;
+          InfoBags[new_type].sizeAll  += size;
+    }
+#endif
+
+    header->type = new_type;
+}
+
+static inline Bag AllocateMasterPointer() {
+  // HOOK: Allocate memory for the master pointer.
+  // Master pointers require one word of memory.
+  void *result = (void *) jl_gc_alloc(JuliaTLS,
+    sizeof(void *), datatype_mptr);
+  return result;
```

Comment:
I was wondering whether it would make more sense for Boehm and Julia GC to store at lest the type and flags (but possibly also the size) inside the master point instead? This way, querying the type of an object requires on indirection less.